### PR TITLE
Improve BinaryDelta error handling & logging

### PR DESCRIPTION
**Pull request by zorgiepoo**
* Fix issue #608. From the framework, the error is logged via SULog. From the tool, the error is logged to stderr. Uses NSError handling for both.
* Move patch version validation for creating a binary diff into the BinaryDelta tool
* Add parameter assertions to createBinaryDelta and handle if -getFileSystemRepresentation:maxLength: fails

### CODE REVIEW

File: Sparkle/SUBinaryDeltaApply.m [hunk 1]
```diff
@@ -25,12 +25,14 @@ static BOOL applyBinaryDeltaToFile(xar_t x, xar_file_t file, NSString *sourceFil
...
 }
 
-int applyBinaryDelta(NSString *source, NSString *destination, NSString *patchFile, BOOL verbose)
+BOOL applyBinaryDelta(NSString *source, NSString *destination, NSString *patchFile, BOOL verbose, NSError * __autoreleasing *error)
 {
     xar_t x = xar_open([patchFile fileSystemRepresentation], READ);
     if (!x) {
-        fprintf(stderr, "Unable to open %s. Giving up.\n", [patchFile fileSystemRepresentation]);
-        return 1;
+        if (error != NULL) {
```
Comment (jakepetroules) [inline Sparkle/SUBinaryDeltaApply.m]: use nil with objc objects not NULL

File: Sparkle/SUBinaryDeltaApply.m [hunk 2]
```diff
@@ -106,13 +114,19 @@ int applyBinaryDelta(NSString *source, NSString *destination, NSString *patchFil
     
     NSString *beforeHash = hashOfTreeWithVersion(source, majorDiffVersion);
     if (!beforeHash) {
-        fprintf(stderr, "\nUnable to calculate hash of tree %s\n", [source fileSystemRepresentation]);
-        return 1;
+        if (verbose) fprintf(stderr, "\n");
```
Comment (jakepetroules) [inline Sparkle/SUBinaryDeltaApply.m]: No one liners

File: Sparkle/SUBinaryDeltaApply.m (hunk 1 above)
Comment (zorgiepoo) [inline Sparkle/SUBinaryDeltaApply.m]: NULL is the correct choice actually. It's a pointer to an obj-c object.

**Comment by pornel** [comment]
LGTM

**Pull request merged**